### PR TITLE
[Don't Merge][Need discussion] First-step toward a working mp on Travis

### PR DESCRIPTION
**Pull request by Dref360**
### Summary
This PR solves some issue when a worker dies and the Pool is not told. (OOM, pkill, etc)

This puts a timeout on the future and notifies the user of the issue.

### Discussion
What should we do for those samples in the case of Sequence? Should we re-queue the task or compute the sample directly or just drop it?

### PR Overview

- [x] This PR requires new unit tests [y/n] (make sure tests are included)
- [ ] This PR requires to update the documentation [y/n] (make sure the docs are up-to-date)
- [x] This PR is backwards compatible [y/n]
- [ ] This PR changes the current API [y/n] (all API changes need to be approved by fchollet)

### CODE REVIEW

**Comment by gabrieldemarmiesse** [comment]
If I understand correctly (I'm by no means an expert in multithreading/multiprocessing), in the implementation proposed in this PR, the batch is dropped and a warning is raised. Is that right?

If it's not too much trouble to implement, I believe the best option is to recompute the sample and raise a warning for those reasons:
* We don't take the risk of breaking things since other keras components might crash/behave in a weird way if the Sequence doesn't yield the right number of batches.
* We keep the current behavior. It's also the expected behavior by the users. 
* Changing the order of the batches can cause users to wonder what is going on (silent unexpected behavior) if they overlook the warning, but this is unlikely to happen in practice.
* If the crash is caused by a sample and is deterministic, the user is interested in the stacktrace. If we recompute it in the main process/thread, the stacktrace will be displayed.

I might be wrong on certain points, as I don't know much about all this.

File: keras/utils/data_utils.py [hunk 1]
```diff
@@ -568,8 +568,10 @@ def _run(self):
                 for i in sequence:
                     if self.stop_signal.is_set():
                         return
+                    future = executor.apply_async(get_index, (self.uid, i))
+                    future.idx = i
                     self.queue.put(
-                        executor.apply_async(get_index, (self.uid, i)), block=True)
+                        future, block=True)
```
Comment (gabrieldemarmiesse) [inline keras/utils/data_utils.py]: I don't think the line needs to be broken.

File: keras/utils/data_utils.py [hunk 2]
```diff
@@ -684,8 +695,18 @@ def get(self):
...
+                try:
+                    future = self.queue.get(block=True)
+                    inputs = future.get(timeout=30)
+                    self.queue.task_done()
+                except mp.TimeoutError:
+                    warnings.warn(
+                        "An input could not be retrieved."
+                        " It could be because a worker has died."
+                        "We do not have any information on the lost sample."
+                            .format(),
```
Comment (gabrieldemarmiesse) [inline keras/utils/data_utils.py]: I don't believe the format is useful.

File: keras/utils/data_utils.py [hunk 3]
```diff
@@ -594,8 +596,17 @@ def get(self):
...
-                inputs = self.queue.get(block=True).get()
-                self.queue.task_done()
+                try:
+                    future = self.queue.get(block=True)
+                    inputs = future.get(timeout=30)
+                    self.queue.task_done()
+                except mp.TimeoutError:
+                    idx = future.idx
+                    warnings.warn(
+                        "The input {} could not be retrieved."
```
Comment (fchollet) [inline keras/utils/data_utils.py]: Nit: use `'` as quote character for consistency.

Same below.

File: keras/utils/data_utils.py (hunk 2 above)
Comment (fchollet) [reply keras/utils/data_utils.py]: +1, maybe you intended to display the index in the warning?

**Comment by Dref360** [comment]
As of now, there is no way to inform `fit_generator` to skip an index. So every step will be delayed.

File: keras/utils/data_utils.py (hunk 2 above)
Comment (Dref360) [reply keras/utils/data_utils.py]: Yeah, but we cannot know the index in a generator.

**Comment by gabrieldemarmiesse** [comment]
Currently, most of the timouts append with CNTK and python 3.6. So fixing the timeout for 3.6 would be a very good first step.

File: tests/keras/utils/data_utils_test.py [hunk 4]
```diff
@@ -391,5 +403,38 @@ def test_finite_generator_enqueuer_processes():
...
+    enqueuer = OrderedEnqueuer(TimeOutSequence([3, 2, 2, 3]),
+                               use_multiprocessing=True)
+    enqueuer.start(3, 10)
+    gen_output = enqueuer.get()
+    with warnings.catch_warnings(record=True) as w:
+        warnings.simplefilter("always")
+        for _ in range(11):
+            next(gen_output)
+    assert "The input {} could not be retrieved.".format(
+        missing_idx) in str(w[-1].message)
```
Comment (gabrieldemarmiesse) [inline tests/keras/utils/data_utils_test.py]: The test will fail if other warnings are being emitted. I would recommend using pytest's utility to check that warnings are being triggered correctly: https://docs.pytest.org/en/latest/warnings.html#warns

File: tests/keras/utils/data_utils_test.py [hunk 5]
```diff
@@ -391,5 +403,38 @@ def test_finite_generator_enqueuer_processes():
...
+
+    enqueuer = GeneratorEnqueuer(create_finite_generator_from_sequence_pcs(
+        TimeOutSequence([3, 2, 2, 3])), use_multiprocessing=True)
+    enqueuer.start(3, 10)
+    gen_output = enqueuer.get()
+    with warnings.catch_warnings(record=True) as w:
+        warnings.simplefilter("always")
+        for _ in range(4 * missing_idx):
+            next(gen_output)
+    assert 'An input could not be retrieved.' in str(w[-1].message)
```
Comment (gabrieldemarmiesse) [inline tests/keras/utils/data_utils_test.py]: Same here.

**Comment by gabrieldemarmiesse** [comment]
@Dref360 I updated the PR.

@fchollet could you review? Thank you.

**Comment by gabrieldemarmiesse** [comment]
Ping @fchollet could you review when you have the time? Thank you.

**Comment by Dref360** [comment]
I think this is LGTM, this would give the user "some" information and we can move forward with using "spawn" when possible on Travis.

**Comment by gabrieldemarmiesse** [comment]
Thanks @Dref360 for working on this!

**Pull request merged**